### PR TITLE
Тягунов -- Лабораторная работа #3

### PR DESCRIPTION
**Pull request by vtyagunov**

### CODE REVIEW

**Comment by vtyagunov** [comment]
@AndrewSmrnv, @KalyanovDO 
Please, review

**Comment by KalyanovDO** [comment]
:+1:

**Comment by AndrewSmrnv** [comment]
👍

File: code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.BufferedWriter;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+
+public class TxtLogger implements ILogger {
+    private static final String DATE_TIME_FORMAT = "yyyy-MM-dd HH:mm:ss";
```
Comment (VitKonevskiy) [inline code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java]: "yyyy-MM-dd HH:mm:ss.SSS" should be more usefull

File: code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+        }
+        bufferedWriter = logWriter;
+    }
+
+    @Override
+    public void log(final String s) {
+        try {
+            bufferedWriter.write(currentLocalDataTime() + " >> " + s);
```
Comment (VitKonevskiy) [inline code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java]: Pls, move the separate to constants

File: code/string-calculator/view/src/main/java/ru/unn/agile/calc/view/StringCalcView.java [hunk 3]
```diff
@@ -32,11 +34,19 @@ void initialize() {
...
+            }
+        });
+
         resultLbl.textProperty().bindBidirectional(viewModel.resultLblProperty());
 
         calcBtn.disableProperty().bindBidirectional(viewModel.calculationDisabledProperty());
 
         calcBtn.setOnAction(e -> viewModel.calculate());
+
+
```
Comment (VitKonevskiy) [inline code/string-calculator/view/src/main/java/ru/unn/agile/calc/view/StringCalcView.java]: redundant lines?

File: code/string-calculator/viewmodel/src/test/java/ru/unn/agile/calc/viewmodel/StringCalcViewModelTests.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+    private void setExpressionAndCalculate(final String expression) {
+        setExpression(expression);
+        calculate();
+    }
+
+    private List<String> getLog() {
+        return viewModel.getLog();
+    }
+
```
Comment (VitKonevskiy) [inline code/string-calculator/viewmodel/src/test/java/ru/unn/agile/calc/viewmodel/StringCalcViewModelTests.java]: redundant line

**Review by VitKonevskiy — CHANGES_REQUESTED** [review]

File: code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java (hunk 1 above)
Comment (vtyagunov) [reply code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java]: done

File: code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java (hunk 2 above)
Comment (vtyagunov) [reply code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java]: done

File: code/string-calculator/view/src/main/java/ru/unn/agile/calc/view/StringCalcView.java (hunk 3 above)
Comment (vtyagunov) [reply code/string-calculator/view/src/main/java/ru/unn/agile/calc/view/StringCalcView.java]: fixed

File: code/string-calculator/viewmodel/src/test/java/ru/unn/agile/calc/viewmodel/StringCalcViewModelTests.java (hunk 4 above)
Comment (vtyagunov) [reply code/string-calculator/viewmodel/src/test/java/ru/unn/agile/calc/viewmodel/StringCalcViewModelTests.java]: fixed

File: code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+        }
+        bufferedWriter = logWriter;
+    }
+
+    @Override
+    public void log(final String s) {
+        try {
+            bufferedWriter.write(currentLocalDataTime() + " " + SEPARATOR_FORMAT + " " + s);
```
Comment (VitKonevskiy) [inline code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java]: Pls, use String.format()

File: code/string-calculator/infrastructure/src/test/java/ru/unn/agile/calc/infrastructure/TxtLoggerTests.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Test
+    public void isLogFileCreated() {
+        BufferedReader reader;
+
+        try {
+            reader = new BufferedReader(new FileReader(FILENAME));
+        } catch (FileNotFoundException exception) {
+            fail("File not found.");
```
Comment (VitKonevskiy) [inline code/string-calculator/infrastructure/src/test/java/ru/unn/agile/calc/infrastructure/TxtLoggerTests.java]: Move to constant, pls

File: code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java [hunk 7]
```diff
@@ -77,4 +129,27 @@ private boolean isCalculationDisabled() {
...
+    public StringProperty logsProperty() {
+        return logs;
+    }
+
+    public final String getLogs() {
+        return logs.get();
+    }
+
+    public final List<String> getLog() {
+        return logger.getLog();
```
Comment (VitKonevskiy) [inline code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java]: What should happend if logger is null? (Default constructor of ViewModel)

**Review by VitKonevskiy — CHANGES_REQUESTED** [review]

File: code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java (hunk 7 above)
Comment (vtyagunov) [reply code/string-calculator/viewmodel/src/main/java/ru/unn/agile/calc/viewmodel/StringCalcViewModel.java]: added NPE check

File: code/string-calculator/infrastructure/src/test/java/ru/unn/agile/calc/infrastructure/TxtLoggerTests.java (hunk 6 above)
Comment (vtyagunov) [reply code/string-calculator/infrastructure/src/test/java/ru/unn/agile/calc/infrastructure/TxtLoggerTests.java]: done

File: code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java (hunk 5 above)
Comment (vtyagunov) [reply code/string-calculator/infrastructure/src/main/java/ru/unn/agile/calc/infrastructure/TxtLogger.java]: done

**Review by VitKonevskiy — APPROVED** [review]

**Pull request merged**